plugins/plugin.go

```diff
@@ -1,119 +1,143 @@
 package plugins
 
 import (
-	"github.com/taskcluster/taskcluster-worker/engine"
+	"io"
+
+	"github.com/taskcluster/taskcluster-worker/engines"
 	"github.com/taskcluster/taskcluster-worker/runtime"
 )
 
-// PluginFactory holds the global state for a plugin.
-//
-// All method on this interface must be thread-safe.
-type PluginFactory interface {
-	// NewPlugin method will be called once for each task. The Plugin instance
-	// returned will be called for each stage in the task execution.
-	//
-	// This method is a called before NewSandboxBuilder(), this is not a good
-	// place to do any operation that may fail as you won't be able to log
-	// anything. This is, however, the place to register things that you wish to
-	// expose to engine and other plugins, such as a log drain.
-	//
-	// Notice, if for some reason the implementor doesn't wish to return a plugin
-	// perhaps runtime.SandboxContextBuilder contains task information for that
-	// disables the plugin, the implementor can safely return an instance of
-	// PluginBase. Such and instance will do absolutely nothing.
-	NewPlugin(builder *runtime.SandboxContextBuilder) Plugin
-}
+// An ExceptionReason specifies the reason a task reached an exception state.
+type ExceptionReason int
+
+// Reasons why a task can reach an exception state. Implementors should be
+// warned that additional entries may be added in the future.
+const (
+	Cancelled ExceptionReason = iota
+	MalformedPayload
+	WorkerShutdown
+)
 
 // Plugin holds the task-specific state for a plugin
 //
 // Each method on this interface represents stage in the task execution and will
-// be called when this stage is reached. The methods are allowed allowed to
+// be called when this stage is reached. The some methods are allowed to
 // take significant amounts of time, as they will run asynchronously.
 //
-// These methods does not have to be thread-safe, we will not call the next
+// These methods does not have to be thread-safe, we will never call the next
 // method, before the previous method has returned.
 //
-// implementors of this interface should be sure to embed PluginBase. This will
+// Implementors of this interface should be sure to embed PluginBase. This will
 // do absolutely nothing, but provide empty implementations for any current and
 // future methods that isn't implemented.
 //
-// If a required feature is unsupport the methods may return a
-// MalformedPayloadError. All other errors are fatal.
+// The methods are called in the order listed here with the exception of
+// Exception() which may be called following any method, and Dispose() which
+// will always be called as a final step allowing you to clean up.
 type Plugin interface {
-	// Prepare will be called in parallel with NewSandboxBuilder(), this is a good
-	// place to start downloading and extracting resources required.
+	PayloadSchema() runtime.CompositeSchema
```

Comment:
`PayloadSchema` returns the `CompositeSchema` that is used to construct the `Payload` property on `PluginOptions`... Hence, it needs to exist at a higher level...